### PR TITLE
Add pseudo-nodes for known services

### LINKED ISSUE #1082: show well known services
Well-known hosts and/or services, such as Amazon RDS, SQS + others TBC should be displayed as a Host node, separate to 'The Internet' node(s)

### PR DESCRIPTION
**Pull request by 2opremio**
Fixes #1082

### CODE REVIEW

**Comment by 2opremio** [comment]
We already show pseudo nodes (with a cloud shape for now), for Dynamo and S3: the two AWS services for which we know reverse DNS lookups lead to an observable pattern.

screen shot 2016-09-14 at 11 09 18 am


Next step is adding a DNS packet inspector in the probes to extend it to IPs without revese lookups.

File: prog/main.go [hunk 1]
```diff
@@ -227,6 +230,10 @@ func main() {
 	flag.StringVar(&flags.probe.weaveAddr, "probe.weave.addr", "127.0.0.1:6784", "IP address & port of the Weave router")
 	flag.StringVar(&flags.probe.weaveHostname, "probe.weave.hostname", app.DefaultHostname, "Hostname to lookup in WeaveDNS")
 
+	// DNS Snooper
+	flag.BoolVar(&flags.probe.dnsSnooperEnabled, "probe.dns-snooper.enabled", true, "Enable DNS Snooper (to identify external services being contacted)")
+	flag.StringVar(&flags.probe.dnsSnooperInterface, "probe.dns-snooper.interface", "any", "Networking interface on which to snoop")
+
```
Comment (rade) [inline prog/main.go]: IMO this is overkill. YAGNI. And we want this 'on' by default.

Comment (rade) [inline prog/main.go]: `eth0` is of no use when connected to a wlan, which many of our developer first-time users will be.

I'd leave it as 'all'.

File: probe/endpoint/reporter.go [hunk 2]
```diff
@@ -193,9 +195,11 @@ func (r *Reporter) makeEndpointNode(namespaceID string, addr string, port uint16
...
 		report.MakeEndpointNodeID(r.hostID, namespaceID, addr, portStr),
 		map[string]string{Addr: addr, Port: portStr})
-	// In case we have a reverse resolution for the IP, we can use it for
-	// the name...
-	if names, err := r.reverseResolver.get(addr); err == nil {
+	names := r.dnsSnooper.CachedNamesForIP(addr)
+	if resolvedNames, err := r.reverseResolver.get(addr); err == nil {
+		names = append(names, resolvedNames...)
+	}
+	if len(names) > 0 {
```
Comment (rade) [inline probe/endpoint/reporter.go]: Stuffing the names we obtained from snooping forward resolutions into the `ReverseDNSNames` set is rather misleading.

So I reckon we should either rename it (but the app will also have to pay attention to the old name, for backward compatibility), or introduce a new set, just for the snooped names.

File: render/process.go [hunk 3]
```diff
@@ -96,8 +96,8 @@ func MapEndpoint2Pseudo(n report.Node, local report.Networks) report.Nodes {
 
 	if ip := net.ParseIP(addr); ip != nil && !local.Contains(ip) {
 		// If the dstNodeAddr is not in a network local to this report, we emit an
-		// internet node
-		node = theInternetNode(n)
+		// external pseudoNode
+		node = toInternetNode(n)
```
Comment (rade) [inline render/process.go]: `externalNode(n)`?

File: render/theinternet.go [hunk 4]
```diff
@@ -2,11 +2,50 @@ package render
...
 import (
 	"net"
+	"regexp"
 
 	"github.com/weaveworks/scope/probe/host"
 	"github.com/weaveworks/scope/report"
 )
 
+var (
+	// ServiceNodeIDPrefix is how the ID all service pseudo nodes begin
```
Comment (rade) [inline render/theinternet.go]: "... the ID *of* all service pseudo nodes..."

File: render/theinternet.go [hunk 5]
```diff
@@ -2,11 +2,50 @@ package render
...
+		regexp.MustCompile(`^.+\.amazonaws\.com$`),
+		regexp.MustCompile(`^.+\.googleapis\.com$`),
+	}
+
+	knownServicesExcluder = []*regexp.Regexp{
+		// We exclude ec2 machines because they are too generic
+		// and having separate nodes for them makes visualizations worse
+		regexp.MustCompile(`^ec2.*\.amazonaws\.com$`),
+	}
+)
```
Comment (rade) [inline render/theinternet.go]: drop the plural. `knownServiceBlah`

Comment (rade) [inline render/theinternet.go]: I wonder whether the excluders should be associated with a matcher, i.e. a matcher becomes a pair of regexps.

Comment (rade) [inline render/theinternet.go]: The matchers/excludersd should really be obtained by the probes. For the moment I'll settle for a 'TODO' and associated issue.

Comment (rade) [inline render/theinternet.go]: ah, but that would make them harder to merge (see next point).

File: prog/main.go (hunk 1 above)
Comment (rade) [inline prog/main.go]: *both* the "enabled" and the interface flags are overkill. Yes, we can think of reasons people may want them. But in practice I just don't think it's going to happen. And if it does then we can always add the configurability then.

File: probe/endpoint/dns_snooper_linux_amd64.go [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+	return s, nil
+}
+
+func newPcapHandle(networkInterface string) (*pcap.Handle, error) {
+	inactive, err := pcap.NewInactiveHandle(networkInterface)
+	if err != nil {
+		return nil, err
+	}
+	defer inactive.CleanUp()
+	if err = inactive.SetPromisc(true); err != nil {
```
Comment (rade) [inline probe/endpoint/dns_snooper_linux_amd64.go]: do we need this?

File: prog/main.go (hunk 1 above)
Comment (2opremio) [inline prog/main.go]: OK, I will (reluctantly :) ) remove them.

File: probe/endpoint/reporter.go (hunk 2 above)
Comment (2opremio) [inline probe/endpoint/reporter.go]: I will add another field.

File: render/process.go (hunk 3 above)
Comment (2opremio) [inline render/process.go]: Sure, I wasn't happy with the name.

File: render/theinternet.go (hunk 5 above)
Comment (2opremio) [inline render/theinternet.go]: I disagree on this one, I think they should be configurable (or extensible) in the app UI.

File: probe/endpoint/dns_snooper_linux_amd64.go (hunk 6 above)
Comment (2opremio) [inline probe/endpoint/dns_snooper_linux_amd64.go]: It works without it (at least in my dev VM), will remove.

**Comment by 2opremio** [comment]
@rade PTAL

File: render/theinternet.go (hunk 5 above)
Comment (rade) [inline render/theinternet.go]: That just ain't going to happen anytime soon. Way too many complications in doing that. It fundamentally changes the nature of the scope app - to date *all* information comes from the probes.

Meanwhile, releasing this feature w/o giving users any way to adjust the matchers/exclusions to their environment is just mean.

Comment (2opremio) [inline render/theinternet.go]: I will add a TODO but not mentioning where it should be configurable.

Comment (2opremio) [inline render/theinternet.go]: Done https://github.com/weaveworks/scope/issues/1876

**Comment by rade** [comment]
needs some squashing. rest LGTM

**Pull request merged**